### PR TITLE
Support NumPy 1.20

### PR DESCRIPTION
**Pull request by stuartarchibald**
This adds support for NumPy 1.20, also raises the minimum
supported version of NumPy to 1.17.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Issues with 1035f18:

Python 3.9 NumPy 1.20:
```
======================================================================
FAIL: test_windowing (numba.tests.test_np_functions.TestNPFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_np_functions.py", line 3748, in test_windowing
    check_window(np_bartlett)
  File "/home/vsts/work/1/s/numba/tests/test_np_functions.py", line 3741, in check_window
    self.assertPreciseEqual(expected, got, prec='double')
  File "/home/vsts/work/1/s/numba/tests/support.py", line 346, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [0.         0.18181818 0.36363636 0.54545455 0.72727273 0.90909091
 0.90909091 0.72727273 0.54545455 0.36363636 0.18181818 0.        ] and [0.         0.18181818 0.36363636 0.54545455 0.72727273 0.90909091
 0.90909091 0.72727273 0.54545455 0.36363636 0.18181818 0.        ]: 0.18181818181818177 != 0.18181818181818182 within 4.037174635000569e-17 delta (5.551115123125783e-17 difference)

======================================================================
FAIL: test_windowing (numba.tests.test_np_functions.TestNPFunctions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/support.py", line 731, in tearDown
    self.memory_leak_teardown()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 705, in memory_leak_teardown
    self.assert_no_memory_leak()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 714, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 12 != 11

```

Python 3.8, NumPy 1.20:
```
======================================================================
FAIL: test_nanquantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_array_reductions.py", line 530, in test_nanquantile_basic
    self.check_percentile_and_quantile(pyfunc, q_upper_bound=1)
  File "/home/vsts/work/1/s/numba/tests/test_array_reductions.py", line 372, in check_percentile_and_quantile
    check(a, q)
  File "/home/vsts/work/1/s/numba/tests/test_array_reductions.py", line 359, in check
    self.assertPreciseEqual(got, expected, abs_tol=abs_tol)
  File "/home/vsts/work/1/s/numba/tests/support.py", line 346, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [       -inf -1.70182786 -0.46359554         nan -0.53228227        -inf
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
        -inf -0.77504848] and [        nan -1.70182786 -0.46359554         nan -0.53228227         nan
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
         nan -0.77504848]: -inf != nan

Stderr:
/home/vsts/miniconda3/envs/azure_ci/lib/python3.8/site-packages/numpy/lib/function_base.py:3961: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
/home/vsts/miniconda3/envs/azure_ci/lib/python3.8/site-packages/numpy/lib/function_base.py:3963: RuntimeWarning: invalid value encountered in add
  lerp_interpolation = asanyarray(add(a, diff_b_a*t, out=out))

======================================================================
FAIL: test_nanquantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/support.py", line 731, in tearDown
    self.memory_leak_teardown()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 705, in memory_leak_teardown
    self.assert_no_memory_leak()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 714, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 69 != 68

Stderr:
/home/vsts/miniconda3/envs/azure_ci/lib/python3.8/site-packages/numpy/lib/function_base.py:3961: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
/home/vsts/miniconda3/envs/azure_ci/lib/python3.8/site-packages/numpy/lib/function_base.py:3963: RuntimeWarning: invalid value encountered in add
  lerp_interpolation = asanyarray(add(a, diff_b_a*t, out=out))

----------------------------------------------------------------------
```

**Comment by stuartarchibald** [comment]
Note: 335f339 is a temporary patch just to get CI working for 32 bit linux, it pulls in NumPy and SciPy from pypi as there's no conda package of NumPy since1.15.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -28,91 +28,92 @@ jobs:
...
         PYTHON: '3.7'
         NUMPY: '1.15'
         CONDA_ENV: azure_ci
-        RUN_COVERAGE: yes
-        RUN_FLAKE8: yes
-        RUN_MYPY: yes
+        BITS32: yes
         TEST_START_INDEX: 2
-      py37_np115_32bit:
+      py37_np117_cov:
```
Comment (sklam) [inline azure-pipelines.yml:38]: Why is the ordering swapped here? From cov first to 32-bit first?

**Review by sklam — COMMENTED** [review]
If we are to revert 335f339 later, we will need to make our own 32-bit numpy packages in the numba channel. I guess we will also need that for buildfarm.

File: azure-pipelines.yml (hunk 1 above)
Comment (stuartarchibald) [reply azure-pipelines.yml:38]: I swapped them as they indexed in vaguely lexicographical order, `py37_np115` < `py37_np117`, however... I'm not sure why this is still NumPy 1.15 for 32 bit and how CI is actually managing to get NumPy 1.17 from pip. Suggest we don't merge this until this is figured out.

**Comment by esc** [comment]
This branch currently has some conflicts with master.

**Comment by esc** [comment]
Build Farm ID: `numba_smoketest_cpu_yaml_31`

**Comment by esc** [comment]
Buildfarm passed on this one.

**Review by sklam — APPROVED** [review]
Thanks for the patch!

**Pull request merged**